presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -362,21 +385,42 @@ private void collectDynamicFilters(StageId stageId, Optional<Set<DynamicFilterId
         return Optional.ofNullable(dynamicFilterContexts.get(queryId).getDynamicFilterSummaries().get(filterId));
     }
 
-    private static TupleDomain<ColumnHandle> translateSummaryToTupleDomain(DynamicFilterId filterId, Domain summary, Multimap<DynamicFilterId, ColumnHandle> sourceColumnHandles)
+    private TupleDomain<ColumnHandle> translateSummaryToTupleDomain(
+            DynamicFilterId filterId,
+            DynamicFilterContext dynamicFilterContext,
+            Multimap<DynamicFilterId, Symbol> symbolsMap,
+            Map<Symbol, ColumnHandle> columnHandles,
+            TypeProvider typeProvider)
     {
-        checkState(sourceColumnHandles.containsKey(filterId), "No source column handles for dynamic filter %s", filterId);
-        return TupleDomain.withColumnDomains(
-                sourceColumnHandles.get(filterId).stream()
-                        .collect(toImmutableMap(handle -> handle, handle -> summary)));
+        Collection<Symbol> probeSymbols = symbolsMap.get(filterId);
+        checkState(probeSymbols != null, "No probe symbols for dynamic filter %s", filterId);
+        Domain summary = dynamicFilterContext.getDynamicFilterSummaries().get(filterId);
+        return TupleDomain.withColumnDomains(probeSymbols.stream()
+                .collect(toImmutableMap(
+                        probeSymbol -> requireNonNull(columnHandles.get(probeSymbol), () -> format("Missing probe column for %s", probeSymbol)),
+                        probeSymbol -> {
+                            Type targetType = typeProvider.get(probeSymbol);
+                            if (!summary.getType().equals(targetType)) {
+                                return DomainCoercer.applySaturatedCasts(metadata, typeOperators, dynamicFilterContext.getSession(), summary, targetType);
+                            }
+                            return summary;
+                        })));
     }
 
-    private static Multimap<DynamicFilterId, ColumnHandle> extractSourceColumnHandles(List<DynamicFilters.Descriptor> dynamicFilters, Map<Symbol, ColumnHandle> columnHandles)
+    private static Multimap<DynamicFilterId, Symbol> extractSourceSymbols(List<DynamicFilters.Descriptor> dynamicFilters, Map<Symbol, ColumnHandle> columnHandles)
     {
         return dynamicFilters.stream()
-                .filter(descriptor -> descriptor.getInput() instanceof SymbolReference)
                 .collect(toImmutableListMultimap(
                         DynamicFilters.Descriptor::getId,
-                        descriptor -> columnHandles.get(Symbol.from(descriptor.getInput()))));
+                        descriptor -> {
```

Comment:
Let's extract this to `io.prestosql.sql.DynamicFilters`. I think it's being used in more than one place

Reply:
done